src/pip/_internal/build_env.py

```diff
@@ -42,29 +64,25 @@ def __init__(self, path: str) -> None:
 
 
 @contextlib.contextmanager
-def _create_standalone_pip() -> Generator[str, None, None]:
-    """Create a "standalone pip" zip file.
+def _create_runnable_pip() -> Generator[str, None, None]:
+    """Create a "pip runner" file.
 
-    The zip file's content is identical to the currently-running pip.
+    The runner file ensures that import for pip happen using the currently-running pip.
     It will be used to install requirements into the build environment.
     """
     source = pathlib.Path(pip_location).resolve().parent
 
-    # Return the current instance if `source` is not a directory. We can't build
-    # a zip from this, and it likely means the instance is already standalone.
+    # Return the current instance if `source` is not a directory. It likely
+    # means that this copy of pip is already standalone.
```

Comment:
With the runner, `pip_location` *will* point to the current source. So this test won't work as expected, and we'll create a second runner hook. Can we do something by looking at pip.__spec__?

In practice, I don't think having multiple runner hooks is a big issue, so I'm OK if we don't worry too much (although maybe note that we've ignored the issue in a comment, to help people looking at this code in future 😉)

Reply:
Well... this check is still useful for the zipapp case.

I'd prefer to defer the reuse of the runner script for a follow-up though. We should be able to make the runner script into a module and call that from the sources directly (I consider inlining code in a string of a module to be a code smell) -- avoiding the need to create a runner script on device in the first place. :)

Reply:
I've filed #11262, which does this FWIW.